pkg/mimirpb/testutil/buffer.go

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package testutil
+
+import (
+	"fmt"
+
+	"go.uber.org/atomic"
+	"google.golang.org/grpc/mem"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// TrackBufferRefCount instruments the WriteRequest's underlying buffer's
+// reference count. It can be inspected with [BufferRefCount].
+func TrackBufferRefCount(wr *mimirpb.WriteRequest) {
+	buf := wr.Buffer()
+	if buf == nil {
+		// Just set some fake buffer. We care about the reference count, not the
+		// data itself.
+		buf = mem.SliceBuffer([]byte("fake data"))
+	}
+	ibuf := &memBufferWithInstrumentedRefCount{Buffer: buf}
+	ibuf.refCount.Add(1) // Match the refCount of buf.Buffer
```

Comment:
"// Match the refCount of buf.Buffer" -- is this necessarily true? If this is test support code, maybe we need to assert it.